File: strategy-game-frontend/src/core/actions/user/userLoginActionImpl.ts
import {AuthData} from "../../../ports/models/authData";
import {UserLoginAction} from "../../../ports/provided/user/userLoginAction";
import {UserApi} from "../../../ports/required/api/userApi";
import {UserStateAccess} from "../../../ports/required/state/userStateAccess";

export class UserLoginActionImpl implements UserLoginAction {

    private readonly userApi: UserApi;
    private readonly userStateAccess: UserStateAccess;

    constructor(userApi: UserApi, userStateAccess: UserStateAccess) {
        this.userApi = userApi;
        this.userStateAccess = userStateAccess;
    }

    perform(email: string, password: string): Promise<void> {
        return this.userApi.login(email, password).then((authData: AuthData) => {
            this.userStateAccess.setAuth(authData.idToken);
        });
    }

}

File: strategy-game-backend/src/main/kotlin/de/ruegnerlukas/strategygame/backend/core/actions/user/UserLoginActionImpl.kt
package de.ruegnerlukas.strategygame.backend.core.actions.user

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import de.ruegnerlukas.strategygame.backend.ports.models.AuthDataExtended
import de.ruegnerlukas.strategygame.backend.ports.provided.user.UserLoginAction
import de.ruegnerlukas.strategygame.backend.ports.required.UserIdentityService

class UserLoginActionImpl(private val userIdentity: UserIdentityService) : UserLoginAction {

    override fun perform(email: String, password: String): Either<UserLoginAction.UserLoginActionError, AuthDataExtended> {
        return when (val result = userIdentity.authenticate(email, password)) {
            is Either.Right -> result.value.right()
            is Either.Left -> when (result.value) {
                UserIdentityService.NotAuthorizedError -> UserLoginAction.NotAuthorizedError.left()
                UserIdentityService.UserNotConfirmedError -> UserLoginAction.UserNotConfirmedError.left()
                UserIdentityService.UserNotFoundError -> UserLoginAction.UserNotFoundError.left()
            }
        }
    }

}